## Test Data
Test data are similar to train data with only one difference. While for training data, data are available until failure for each engine, for test data, it is available for an arbitrary number of cycles. Using the given test data, out goal is to predict RUL. For test data, true RUL values are also available so that we can compare the performance of our prediction algorithm.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.metrics import confusion_matrix, accuracy_score

### The given test data is not in the form of training data we created, so now we need to convert that to the same format.

In [2]:
raw_data = pd.read_csv('test_FD001.txt', sep=' ', header = None)
raw_data = raw_data.drop([26,27], axis='columns')
raw_data.columns = ['ID', 'Cycle', 'OpSet1', 'OpSet2', 'OpSet3', 'SensorMeasure1', 'SensorMeasure2', 'SensorMeasure3', 'SensorMeasure4', 'SensorMeasure5', 'SensorMeasure6', 'SensorMeasure7', 'SensorMeasure8', 'SensorMeasure9', 'SensorMeasure10', 'SensorMeasure11', 'SensorMeasure12', 'SensorMeasure13', 'SensorMeasure14', 'SensorMeasure15', 'SensorMeasure16', 'SensorMeasure17', 'SensorMeasure18', 'SensorMeasure19', 'SensorMeasure20', 'SensorMeasure21']


In [3]:
raw_data

ID  Cycle  OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0        1      1  0.0023  0.0003   100.0          518.67          643.02   
1        1      2 -0.0027 -0.0003   100.0          518.67          641.71   
2        1      3  0.0003  0.0001   100.0          518.67          642.46   
3        1      4  0.0042  0.0000   100.0          518.67          642.44   
4        1      5  0.0014  0.0000   100.0          518.67          642.51   
...    ...    ...     ...     ...     ...             ...             ...   
13091  100    194  0.0049  0.0000   100.0          518.67          643.24   
13092  100    195 -0.0011 -0.0001   100.0          518.67          643.22   
13093  100    196 -0.0006 -0.0003   100.0          518.67          643.44   
13094  100    197 -0.0038  0.0001   100.0          518.67          643.26   
13095  100    198  0.0013  0.0003   100.0          518.67          642.95   

       SensorMeasure3  SensorMeasure4  SensorMeasure5  ...  SensorMeasure12  \
0             1585.29         1398.21           14.62  ...           521.72   
1             1588.45         1395.42           14.62  ...           522.16   
2             1586.94         1401.34           14.62  ...           521.97   
3             1584.12         1406.42           14.62  ...           521.38   
4             1587.19         1401.92           14.62  ...           522.15   
...               ...             ...             ...  ...              ...   
13091         1599.45         1415.79           14.62  ...           520.69   
13092         1595.69         1422.05           14.62  ...           521.05   
13093         1593.15         1406.82           14.62  ...           521.18   
13094         1594.99         1419.36           14.62  ...           521.33   
13095         1601.62         1424.99           14.62  ...           521.07   

       SensorMeasure13  SensorMeasure14  SensorMeasure15  SensorMeasure16  \
0              2388.03          8125.55           8.4052             0.03   
1              2388.06          8139.62           8.3803             0.03   
2              2388.03          8130.10           8.4441             0.03   
3              2388.05          8132.90           8.3917             0.03   
4              2388.03          8129.54           8.4031             0.03   
...                ...              ...              ...              ...   
13091          2388.00          8213.28           8.4715             0.03   
13092          2388.09          8210.85           8.4512             0.03   
13093          2388.04          8217.24           8.4569             0.03   
13094          2388.08          8220.48           8.4711             0.03   
13095          2388.05          8214.64           8.4903             0.03   

       SensorMeasure17  SensorMeasure18  SensorMeasure19  SensorMeasure20  \
0                  392             2388            100.0            38.86   
1                  393             2388            100.0            39.02   
2                  393             2388            100.0            39.08   
3                  391             2388            100.0            39.00   
4                  390             2388            100.0            38.99   
...                ...              ...              ...              ...   
13091              394             2388            100.0            38.65   
13092              395             2388            100.0            38.57   
13093              395             2388            100.0            38.62   
13094              395             2388            100.0            38.66   
13095              396             2388            100.0            38.70   

       SensorMeasure21  
0              23.3735  
1              23.3916  
2              23.4166  
3              23.3737  
4              23.4130  
...                ...  
13091          23.1974  
13092          23.2771  
13093          23.2051  
13094          23.2699  
13095    

In [4]:
given_no_of_cycles=[]
for i in range (1,len(raw_data['ID'].unique())+1):
    given_no_of_cycles.append(len(raw_data[raw_data['ID']==i]))
    print(len(raw_data[raw_data['ID']==i]))
    
given_no_of_cycles=np.array(given_no_of_cycles).reshape([cycle_ran_after.shape[0],1])    

31
49
126
106
98
105
160
166
55
192
83
217
195
46
76
113
165
133
135
184
148
39
130
186
48
76
140
158
171
143
196
145
50
203
198
126
121
125
37
133
123
156
172
54
152
146
73
78
303
74
144
189
164
121
113
136
160
176
94
147
159
232
155
168
71
147
71
187
54
152
68
131
112
137
88
205
162
72
101
133
213
162
73
172
34
110
56
68
177
146
234
150
244
133
89
97
134
121
97
198


NameError: name 'cycle_ran_after' is not defined

We can see that data for 31 cycles are available for engine 1 in test set. Data for 49 cycles are available for engine 2, and so on. Using the available data our goal is to predict RUL for each engine.

In [5]:
len(raw_data['ID'].unique())

100

In [6]:
cycle_ran_after = pd.read_csv('RUL_FD001.txt', sep=' ', header = None)
cycle_ran_after=cycle_ran_after.drop([1], axis='columns')
cycle_ran_after = np.array(cycle_ran_after)
cycle_ran_after

array([[112],
       [ 98],
       [ 69],
       [ 82],
       [ 91],
       [ 93],
       [ 91],
       [ 95],
       [111],
       [ 96],
       [ 97],
       [124],
       [ 95],
       [107],
       [ 83],
       [ 84],
       [ 50],
       [ 28],
       [ 87],
       [ 16],
       [ 57],
       [111],
       [113],
       [ 20],
       [145],
       [119],
       [ 66],
       [ 97],
       [ 90],
       [115],
       [  8],
       [ 48],
       [106],
       [  7],
       [ 11],
       [ 19],
       [ 21],
       [ 50],
       [142],
       [ 28],
       [ 18],
       [ 10],
       [ 59],
       [109],
       [114],
       [ 47],
       [135],
       [ 92],
       [ 21],
       [ 79],
       [114],
       [ 29],
       [ 26],
       [ 97],
       [137],
       [ 15],
       [103],
       [ 37],
       [114],
       [100],
       [ 21],
       [ 54],
       [ 72],
       [ 28],
       [128],
       [ 14],
       [ 77],
       [  8],
       [121],
       [ 94],
       [118],
      

An RUL value of 112 for engine 1 in test set means that it failed 112 cycles after 31st cycle. So engine 1 ran for (31+112) cycles before it failed. Similarly engine two failed 98 cycles after 49th cycle. So engine 2 ran for (49 + 98) cycles in total before it failed. Similar calculation can be done for other engines in test set.

# Making EOL column


In [7]:
EOL=[]
i=0

for i in raw_data['ID']:
        EOL.append(given_no_of_cycles[i-1]+cycle_ran_after[i-1] ) 
        
EOL = np.array(EOL).reshape([len(EOL),1])         

In [8]:
EOL.shape

(13096, 1)

In [9]:
# Add the 'EOL' column to 'raw_data'
raw_data['EOL']=EOL
raw_data

ID  Cycle  OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0        1      1  0.0023  0.0003   100.0          518.67          643.02   
1        1      2 -0.0027 -0.0003   100.0          518.67          641.71   
2        1      3  0.0003  0.0001   100.0          518.67          642.46   
3        1      4  0.0042  0.0000   100.0          518.67          642.44   
4        1      5  0.0014  0.0000   100.0          518.67          642.51   
...    ...    ...     ...     ...     ...             ...             ...   
13091  100    194  0.0049  0.0000   100.0          518.67          643.24   
13092  100    195 -0.0011 -0.0001   100.0          518.67          643.22   
13093  100    196 -0.0006 -0.0003   100.0          518.67          643.44   
13094  100    197 -0.0038  0.0001   100.0          518.67          643.26   
13095  100    198  0.0013  0.0003   100.0          518.67          642.95   

       SensorMeasure3  SensorMeasure4  SensorMeasure5  ...  SensorMeasure13  \
0             1585.29         1398.21           14.62  ...          2388.03   
1             1588.45         1395.42           14.62  ...          2388.06   
2             1586.94         1401.34           14.62  ...          2388.03   
3             1584.12         1406.42           14.62  ...          2388.05   
4             1587.19         1401.92           14.62  ...          2388.03   
...               ...             ...             ...  ...              ...   
13091         1599.45         1415.79           14.62  ...          2388.00   
13092         1595.69         1422.05           14.62  ...          2388.09   
13093         1593.15         1406.82           14.62  ...          2388.04   
13094         1594.99         1419.36           14.62  ...          2388.08   
13095         1601.62         1424.99           14.62  ...          2388.05   

       SensorMeasure14  SensorMeasure15  SensorMeasure16  SensorMeasure17  \
0              8125.55           8.4052             0.03              392   
1              8139.62           8.3803             0.03              393   
2              8130.10           8.4441             0.03              393   
3              8132.90           8.3917             0.03              391   
4              8129.54           8.4031             0.03              390   
...                ...              ...              ...              ...   
13091          8213.28           8.4715             0.03              394   
13092          8210.85           8.4512             0.03              395   
13093          8217.24           8.4569             0.03              395   
13094          8220.48           8.4711             0.03              395   
13095          8214.64           8.4903             0.03              396   

       SensorMeasure18  SensorMeasure19  SensorMeasure20  SensorMeasure21  EOL  
0                 2388            100.0            38.86          23.3735  143  
1                 2388            100.0            39.02          23.3916  143  
2                 2388            100.0            39.08          23.4166  143  
3                 2388            100.0            39.00          23.3737  143  
4                 2388            100.0            38.99          23.4130  143  
...                ...              ...              ...              ...  ...  
13091             2388            100.0            38.65          23.1974  218  
13092             2388            100.0            38.57          23.2771  218  
13093             2388            100.0            38.62          23.2051  218  
13094             2388            100.0            38.66          23.2699  218  
13095             2388            100.0            38.70          23.1855  218  

[13096 rows x 27 columns]

In [10]:
raw_data["LR"] = raw_data["Cycle"].div(raw_data["EOL"])
# Drop ID and EOL
raw_data=raw_data.drop(columns=['ID','EOL'])
raw_data

Cycle  OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0          1  0.0023  0.0003   100.0          518.67          643.02   
1          2 -0.0027 -0.0003   100.0          518.67          641.71   
2          3  0.0003  0.0001   100.0          518.67          642.46   
3          4  0.0042  0.0000   100.0          518.67          642.44   
4          5  0.0014  0.0000   100.0          518.67          642.51   
...      ...     ...     ...     ...             ...             ...   
13091    194  0.0049  0.0000   100.0          518.67          643.24   
13092    195 -0.0011 -0.0001   100.0          518.67          643.22   
13093    196 -0.0006 -0.0003   100.0          518.67          643.44   
13094    197 -0.0038  0.0001   100.0          518.67          643.26   
13095    198  0.0013  0.0003   100.0          518.67          642.95   

       SensorMeasure3  SensorMeasure4  SensorMeasure5  SensorMeasure6  ...  \
0             1585.29         1398.21           14.62           21.61  ...   
1             1588.45         1395.42           14.62           21.61  ...   
2             1586.94         1401.34           14.62           21.61  ...   
3             1584.12         1406.42           14.62           21.61  ...   
4             1587.19         1401.92           14.62           21.61  ...   
...               ...             ...             ...             ...  ...   
13091         1599.45         1415.79           14.62           21.61  ...   
13092         1595.69         1422.05           14.62           21.61  ...   
13093         1593.15         1406.82           14.62           21.61  ...   
13094         1594.99         1419.36           14.62           21.61  ...   
13095         1601.62         1424.99           14.62           21.61  ...   

       SensorMeasure13  SensorMeasure14  SensorMeasure15  SensorMeasure16  \
0              2388.03          8125.55           8.4052             0.03   
1              2388.06          8139.62           8.3803             0.03   
2              2388.03          8130.10           8.4441             0.03   
3              2388.05          8132.90           8.3917             0.03   
4              2388.03          8129.54           8.4031             0.03   
...                ...              ...              ...              ...   
13091          2388.00          8213.28           8.4715             0.03   
13092          2388.09          8210.85           8.4512             0.03   
13093          2388.04          8217.24           8.4569             0.03   
13094          2388.08          8220.48           8.4711             0.03   
13095          2388.05          8214.64           8.4903             0.03   

       SensorMeasure17  SensorMeasure18  SensorMeasure19  SensorMeasure20  \
0                  392             2388            100.0            38.86   
1                  393             2388            100.0            39.02   
2                  393             2388            100.0            39.08   
3                  391             2388            100.0            39.00   
4                  390             2388            100.0            38.99   
...                ...              ...              ...              ...   
13091              394             2388            100.0            38.65   
13092              395             2388            100.0            38.57   
13093              395             2388            100.0            38.62   
13094              395             2388            100.0            38.66   
13095              396             2388            100.0            38.70   

       SensorMeasure21        LR  
0              23.3735  0.006993  
1              23.3916  0.013986  
2              23.4166  0.020979  
3              23.3737  0.027972  
4              23.4130  0.034965  
...                ...       ...  
13091          23.1974  0.889908  
13092          23.2771  0.894495  
13093          23.2051  0.899083  
13094

In [11]:
labels=[]
for i in range (0,len(raw_data)):
    if np.array(raw_data["LR"])[i] <= 0.6:
        labels.append(0)
    elif np.array(raw_data["LR"])[i] <= 0.8 :
        labels.append(1)   
    else :
        labels.append(2)  
        
raw_data["labels"]=labels
raw_data  = raw_data.drop(columns=['LR'])   
raw_data

Cycle  OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0          1  0.0023  0.0003   100.0          518.67          643.02   
1          2 -0.0027 -0.0003   100.0          518.67          641.71   
2          3  0.0003  0.0001   100.0          518.67          642.46   
3          4  0.0042  0.0000   100.0          518.67          642.44   
4          5  0.0014  0.0000   100.0          518.67          642.51   
...      ...     ...     ...     ...             ...             ...   
13091    194  0.0049  0.0000   100.0          518.67          643.24   
13092    195 -0.0011 -0.0001   100.0          518.67          643.22   
13093    196 -0.0006 -0.0003   100.0          518.67          643.44   
13094    197 -0.0038  0.0001   100.0          518.67          643.26   
13095    198  0.0013  0.0003   100.0          518.67          642.95   

       SensorMeasure3  SensorMeasure4  SensorMeasure5  SensorMeasure6  ...  \
0             1585.29         1398.21           14.62           21.61  ...   
1             1588.45         1395.42           14.62           21.61  ...   
2             1586.94         1401.34           14.62           21.61  ...   
3             1584.12         1406.42           14.62           21.61  ...   
4             1587.19         1401.92           14.62           21.61  ...   
...               ...             ...             ...             ...  ...   
13091         1599.45         1415.79           14.62           21.61  ...   
13092         1595.69         1422.05           14.62           21.61  ...   
13093         1593.15         1406.82           14.62           21.61  ...   
13094         1594.99         1419.36           14.62           21.61  ...   
13095         1601.62         1424.99           14.62           21.61  ...   

       SensorMeasure13  SensorMeasure14  SensorMeasure15  SensorMeasure16  \
0              2388.03          8125.55           8.4052             0.03   
1              2388.06          8139.62           8.3803             0.03   
2              2388.03          8130.10           8.4441             0.03   
3              2388.05          8132.90           8.3917             0.03   
4              2388.03          8129.54           8.4031             0.03   
...                ...              ...              ...              ...   
13091          2388.00          8213.28           8.4715             0.03   
13092          2388.09          8210.85           8.4512             0.03   
13093          2388.04          8217.24           8.4569             0.03   
13094          2388.08          8220.48           8.4711             0.03   
13095          2388.05          8214.64           8.4903             0.03   

       SensorMeasure17  SensorMeasure18  SensorMeasure19  SensorMeasure20  \
0                  392             2388            100.0            38.86   
1                  393             2388            100.0            39.02   
2                  393             2388            100.0            39.08   
3                  391             2388            100.0            39.00   
4                  390             2388            100.0            38.99   
...                ...              ...              ...              ...   
13091              394             2388            100.0            38.65   
13092              395             2388            100.0            38.57   
13093              395             2388            100.0            38.62   
13094              395             2388            100.0            38.66   
13095              396             2388            100.0            38.70   

       SensorMeasure21  labels  
0              23.3735       0  
1              23.3916       0  
2              23.4166       0  
3              23.3737       0  
4              23.4130       0  
...                ...     ...  
13091          23.1974       2  
13092          23.2771       2  
13093          23.2051       2  
13094          23.2699   

In [69]:
#raw_data.to_csv('Test_classification_4.csv', index=None)